'Dockerize Python app': CI/CD pipeline task completion

rbelshevitz opened the pull request:
A working workflow which do the following:

- packs (builds) a python app into an image, incl. downloading python requirements from a private PyPi-style repo;
- perform login on private docker registry;
- pushes a docker image to that private docker registry;
- runs a container nested from an image build at previous step;
- verifies and stops the container.

rbelshevitz committed:
Create README.md

rbelshevitz committed:
Create wallet.py

rbelshevitz committed:
Create Dockerfile

A Dockerfile to package our small python app.

rbelshevitz committed:
Delete wallet.py

rbelshevitz committed:
This Dockerfile is ok for building image locally.

rbelshevitz committed:
Merge branch 'main' of https://github.com/rbalashevich/python-wallet-api into main

rbelshevitz committed:
Container --name added to docker run

Update docker-image.yml

Image tag fix

Image tag fix #2

docker ps & docker stop added

Update Dockerfile

Job: build a docker image

Secrets removed

env var from github secret added

Added var from secret

Commented string fixed

Secret added

Update docker-image.yml

A skeleton created

Update docker-image-push.yml

Update docker-image-push.yml

Update docker-image-push.yml

Update docker-image-push.yml

Update docker-image.yml

docker run ...

Update docker-image.yml

Docker Registry Login

Modified build and run jobs

Update docker-image.yml

United job

Delete docker-image-push.yml

Update docker-image.yml

antweiss reviewed (changes requested):
Kudos!
This does what is requested.
I like the flow - it's orderly and clear.

Regarding pip.conf - I agree this isn't the solution. You're storing credentials in git and even worse - you're leaving them in the docker image where they aren't needed.

Better options: 

1. put the whole of pip.conf in a github secret and only write it to a file before running docker build (and remove them from the image after pip install is done)
2. pass credentials from a secret to Docker as build arguments (see [here](https://vsupalov.com/docker-arg-env-variable-guide/) and then pass them  directly to pip install : `pip install myWallet -i https://user:password@playtikatraining.jfrog.io/artifactory/api/pypi/pypi-local/simple`
3. you can also install packages to a directlry outside of Docker image E.g: [pip install -t](https://pip.pypa.io/en/latest/cli/pip_install/#cmdoption-t) and then copy them into the images.

Please try one of these approaches. 
the important thing is to avoid keeping creds in the repo.

antweiss commented:
Also - next time please set me the reviewer of your PR - this way you don't even need to send me a mail.

antweiss commented:
And once again - please squash your commits!

rbelshevitz committed:
We do not want put our pip secrets into image!

Update Dockerfile

ARG change to index-url

Update Dockerfile

Update Dockerfile

Update docker-image.yml

Update Dockerfile

Update Dockerfile

Update Dockerfile

Update Dockerfile

Update Dockerfile

Update Dockerfile

Update Dockerfile

Update Dockerfile

Update Dockerfile

Update Dockerfile

Update Dockerfile

Update Dockerfile

Dockerfile for 2-stage build + workflow updated for build argument

Delete pip.conf

rbelshevitz commented:
Here are my changes regarding our so called 'pip.conf secrets workaround'. A combination of options 2 and 3 did the trick. :)

rbelshevitz committed:
k3d + deploy app + check pods

KinD cluster

Create KinD cluster

Create KinD based kubernetes cluster

KinD cluster

Create deployment.yml

Deploy to KinD added

KinD cluster... deployment step disabled

kubectl apply -f $HOME/deployment.yml
  shell: /usr/bin/bash -e {0}
error: the path "/home/runner/deployment.yml" does not exist

Where should I put a file for interaction with kubectl?... How to deal with it? Please explain me. Every control plane warm-up takes more than 5 minutes. :(

KinD cluster: added repo checkout

k3d cluster

k3s action by nolar

Update docker-image.yml

ls .

k3s cluster + deploy + check pod

k3d + deploy + test pod

Some fixes

rbelshevitz committed:
k3d + deploy app + check pods

KinD cluster

Create KinD cluster

Create KinD based kubernetes cluster

KinD cluster

Create deployment.yml

Deploy to KinD added

KinD cluster... deployment step disabled

kubectl apply -f $HOME/deployment.yml
  shell: /usr/bin/bash -e {0}
error: the path "/home/runner/deployment.yml" does not exist

Where should I put a file for interaction with kubectl?... How to deal with it? Please explain me. Every control plane warm-up takes more than 5 minutes. :(

KinD cluster: added repo checkout

k3d cluster

k3s action by nolar

Update docker-image.yml

ls .

k3s cluster + deploy + check pod

k3d + deploy + test pod

rbelshevitz commented:
Update:  1) KinD cluster warms up on every job run, it takes more than 5 minutes for control plane etc. It is annoying. But it starts ok.  2)  kubectl apply -f $HOME/deployment.yml
  shell: /usr/bin/bash -e {0}
error: the path "/home/runner/deployment.yml" does not exist
??? 

Update: talked about, thanks @antweiss .

rbelshevitz commented:
Update:    - uses: actions/checkout@v2 didn't help.

antweiss commented:
@rbalashevich well, check your path (you're still using $HOME) 
Also you can run `ls` in the step to output the file locations, look around and find the files you need

rbelshevitz commented:
Ok, let's talk nolar's k3d action does what we expect, but how to check pod status after one or more minutes? In general, though, it works. @antweiss

antweiss commented on deployment.yml:
```diff
@@ -0,0 +1,39 @@
+    selector:
+        matchLabels:
+            app: wallet-api
+    replicas: 2
```
your deployment currently runs 2 replicas. this wasn't required. please change to 1. this will also make it easier for you to check for running pods.

antweiss commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,84 @@
+          ls .
+          kubectl apply -f ./deployment.yml
+          kubectl get pods
+          echo "Waiting for pods... 1 min..." && sleep 60
```
Sleeping for an arbitrary timeout is not an optimal solution. It would be better to actually check the pod's status in a loop. And limit that loop to a number of iterations whereas each iteration sleeps for a longer period (exponential backoff) until it gives up.

rbelshevitz replied:
What kind of loop you mean?

antweiss commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,84 @@
+          kubectl apply -f ./deployment.yml
+          kubectl get pods
+          echo "Waiting for pods... 1 min..." && sleep 60
+          kubectl get pod -l app=wallet-api -ojson | jq .status.phase 
```
this doesn't actually check for pod's status. it only outputs it to the console. and in your case it's 'null' - please actually check for pod's status. If it's not Running after 2 minutes - fail the workflow.

rbelshevitz replied:
I'll try with another approach. With 'while' loop querying STATUS.

antweiss commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,84 @@
+         name: 6 Stop the container 
+         run: docker stop python-wallet-api
+         
+  # kind-create-kind-cluster:
```
Please remove commented-out code.

antweiss commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+WORKDIR /home/appuser
+USER appuser
+COPY --chown=appuser:appuser requirements.txt requirements.txt
+RUN python -m pip install --upgrade pip
```
It's better to chain Dockerfile RUN instructions with '&&' so we don't create too many layers- e.g:
```
RUN pip install myWallet==0.2 && \
 pip install --user -r requirements.txt && \
 rm /home/appuser/requirements.txt
```

rbelshevitz replied:
I know about layers, sorry, just forgot to chain it.

antweiss reviewed (changes requested):
See my comments:

rbelshevitz committed:
Some fixes

rbelshevitz committed:
While loop added for waiting pod's status.

rbelshevitz committed:
Merge branch 'main' of https://github.com/rbalashevich/python-wallet-api into main

rbelshevitz commented:
Pod's STATUS check added. But the pod seems to not be ready after 2 minutes. I have no idea, why.

antweiss commented:
Well, did you try deploying to your own cluster? Does it run ok there?
Try to run `kubectl describe pod` in the workflow - this may shed some light.

antweiss commented:
It would be helpful to actually print out the status of the pod on each cycle -to see what's happening.

rbelshevitz committed:
k3d: kubectl describe pods added

rbelshevitz committed:
k3s: imagePullPolicy fix in deployment

rbelshevitz committed:
Fixed docker image path in deployment

rbelshevitz committed:
k3s: credentials fix

rbelshevitz committed:
k3s: credentials fix / image pulling fix

k3s: credentials fix / image pulling fix

k3s: credentials fix / image pulling fix #2

Path where to save image.

k3s: credentials fix / image pulling fix #3

k3s: credentials fix / image pulling fix #4

k3s: credentials fix / image pulling fix #6

k3s: credentials fix / image pulling fix #7

k3s: credentials fix / image pulling fix #8

k3s: credentials fix / image pulling fix #9

rbelshevitz commented:
Update: implemented a massive workaround related to lack of 'imagePullSecrets' parameter support in this version (or, maybe, at all!) k3d. ( https://github.com/rancher/k3d/issues/228 ) ... I did a trick to import pulled image into k3d, from a 'local' directory... Also, it is good to keep in mind a so called 'pull policy'. Details are here:
https://k3d.io/usage/commands/k3d_image_import/

A useful article (not about k3d itself, but kinda good): https://iximiuz.com/en/posts/kubernetes-kind-load-docker-image/

antweiss commented:
hi @rbalashevich good work and research on k3d image import. It actually works. Now regarding imagePullSecrets support in k3s - I think you may've gotten it wrong. 
 if I create a secret like this:
`kubectl create secret docker-registry artifactory --docker-server=https://playtikatraining.jfrog.io --docker-username=candidate --docker-password=<redacted> --docker-email=devops@otomato.io`
the following deployment works fine:

```
apiVersion: apps/v1
kind: Deployment
metadata:
    name: wallet-api
spec:
    selector:
        matchLabels:
            app: wallet-api
    replicas: 1
    template:
        metadata:
            labels:
                app: wallet-api
        spec:
            imagePullSecrets:
            - name: artifactory
            containers:
                - name: wallet-api
                  imagePullPolicy: Always
                  image: playtikatraining.jfrog.io/default-docker-local/python-wallet-api
                  resources:
                      limits:
                          cpu: 50m
                          memory: 100Mi
                      requests:
                          cpu: 25m
                          memory: 10Mi
                  ports:
                      - containerPort: 5080
                    ```

antweiss commented:
the error you received was: `ValidationError(Deployment.spec.template.spec.containers[0]): unknown field "imagePullSecrets" in io.k8s.api.core.v1.Container;` 
and that's correct - imagePullSecrets is a field of pod.v1.spec : `kubectl explain pod.spec.imagePullSecrets`

rbelshevitz commented:
Understood thanks. Well, firstly, this is completely, not at all obvious, and secondly, I'm not the only one who faced such a problem. Judging by other threads on the GitHub forums.

kuzishai commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+USER appuser
+COPY --chown=appuser:appuser requirements.txt requirements.txt
+RUN python -m pip install --upgrade pip && \
+    pip install myWallet==0.2  && \
```
is it neccessary to install myWallet==0.2 as root? you have the dependancy in requirements.txt

rbelshevitz replied:
As appuser. Not root.

kuzishai replied:
Interesting question as logs shows that `Requirement already satisfied: myWallet==0.2 in /usr/local/lib/python3.8/site-packages` , I was just curious if the extra step is necessary - I might be wrong but will investigate and will let you know

rbelshevitz replied:
It is obvious if we installed that package before others [available globally in world PyPi]. See the trick with a 'pip.conf' file substitution.

rbelshevitz commented:
'imagePullSecrets' shouldn't be under container definitions, but it should be upper and should be indented in other way. Ok, I understood this nuance. Thanks, will keep this in mind for the next time.

kuzishai commented on deployment.yml:
```diff
@@ -0,0 +1,41 @@
+spec:
+    type: NodePort
+    selector:
+        app: example
```
is it correct?

rbelshevitz replied:
No. There was a typo.

rbelshevitz committed:
Typo fix added

rbalashevich commented:
If, in whatever times, Google and StackOverflow websites would close, your life would change dramatically, @mondbev1 ...

kuzishai commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,77 @@
+          k3d image import /tmp/python-wallet-api.tar
+          kubectl apply -f ./deployment.yml
+          kubectl get pods
+          while [[ $(kubectl get pods -l app=wallet-api -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}') != "True" ]]; do echo "waiting for pod ..." && sleep 15 && kubectl describe pod wallet-api; done
```
I think the extra . here is unnecessary, 
jsonpath={..status.conditions[?(@.type=="Ready")].status}
Also, what happens if the while test fails?
It seems to work, though. great job!

rbelshevitz replied:
Thank you, @mondbev1 .  This result is good anyway.

rbelshevitz replied:
> I think the extra . here is unnecessary, jsonpath={..status.conditions[?(@.type=="Ready")].status} Also, what happens if the while test fails? It seems to work, though. great job!

Just for querying something useful with JSON, nothing else.

antweiss replied:
Roman. I actually understand what @mondbev1 is asking here - if the test fails - i.e if the pod doesn't become ready - the workflow should fail too. Otherwise why test this?

rbelshevitz replied:
Yes, exactly! Got importance of this fact, later.

rbelshevitz committed:
Deploy to EKS

Deploy to EKS #1

Deploy to EKS #2

Deploy to EKS #3

Deploy to EKS #4

Deploy to EKS #5

Docker to EKS #6

Docker to EKS #7

Docker to EKS #8

Docker to EKS #9

Docker to EKS #10

Docker to EKS #11

Docker to EKS #12

Deploy to EKS #13

Deploy to EKS #14

Deploy to EKS #15

Deploy to EKS #15

Deploy to EKS #16

Deploy to EKS #17

Deploy to EKS #18

Deploy to EKS #19

Deploy to EKS #20

Deploy to EKS #21

Deploy to EKS #22

Deploy to EKS #23

Deploy to EKS #24

rbelshevitz committed:
Deploy to EKS: fixes

Deploy to EKS

Deploy to EKS #1

Deploy to EKS #2

Deploy to EKS #3

Deploy to EKS #4

Deploy to EKS #5

Docker to EKS #6

Docker to EKS #7

Docker to EKS #8

Docker to EKS #9

Docker to EKS #10

Docker to EKS #11

Docker to EKS #12

Deploy to EKS #13

Deploy to EKS #14

Deploy to EKS #15

Deploy to EKS #15

Deploy to EKS #16

Deploy to EKS #17

Deploy to EKS #18

Deploy to EKS #19

Deploy to EKS #20

Deploy to EKS #21

Deploy to EKS #22

Deploy to EKS #23

Deploy to EKS #24

Deploy to EKS

Update deployment-eks.yml

Deploy to EKS: fixed spec

rbelshevitz committed:
Deploy to EKS

rbelshevitz committed:
Update deployment-eks.yml

rbelshevitz committed:
Deploy to EKS: fixed spec

rbelshevitz committed:
Deploy to EKS: fixed secret creation

rbelshevitz committed:
Deploy to EKS: fixes

rbelshevitz committed:
Deploy to EKS: fixes

rbelshevitz committed:
Deploy to EKS: fixes

Deploy to EKS: fixes

rbelshevitz commented:
The deployment flow to EKS is working. Sorry, not all commits are 'squashable'.  Samples from GH Marketplace are all a total mess; I implemented all steps separately. 

What I did not understand clear, is how to deal with kubeconfig if we have no awscli utility. Well, it may be another pood of salt, especially with a base64- encoded string. But for now, awscli do the trick.

antweiss commented on .github/workflows/docker-image-eks.yml:
```diff
@@ -0,0 +1,65 @@
+          kubectl get pods
+          while [[ $(kubectl get pods -l app=wallet-api -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}') != "True" ]]; do echo "waiting for pod ..." && sleep 15 && kubectl describe pod wallet-api; done
+
+    # just for memo:
```
Great, I liked this piece of documentation. It could also be scripted in the workflow - i.e - check if the secret exists and if it doesn't - create it based on Github secrets. It's not too complex to implement and  always better to automate than to document.

rbelshevitz committed:
Build, then to k3s (test), then to EKS (prod)

Build, then to k3s (test), then to EKS (prod)

Build, then to k3s (test), then to EKS (prod)

Build, then to k3s (test), then to EKS (prod)

Build, then to k3s (test), then to EKS (prod)

Build, then to k3s (test), then to EKS (prod)

Build, then to k3s (test), then to EKS (prod)

Build, then to k3d (test), then to EKS (prod)

Build, then to k3s (test), then to EKS (prod)

Build, then to k3s (test), then to EKS (prod)

rbelshevitz commented:
Update: I've implemented some checks on so-called real life scenario, in which we move forward to 'prod' (EKS) only, if we're ok on testing (k3s). If not, the flow will not move forward. I hope that a sequence is right.
